### ISSUE TITLE
nodogsplash not starting - dependency out of date, unable to upgrade further

### ISSUE THREAD

**Issue by Kosa103** [nodogsplash/nodogsplash#526]
I followed the provided steps here to make nodogsplash work with raspap:
https://github.com/billz/raspap-webgui/wiki/Captive-portal-setup

However upon completing the service didn't start. Tried to manually start it and I got this:

pi@raspberrypi:~ $ sudo nodogsplash
[5][Sun Feb 16 17:04:38 2020][2142](src/main.c:541) Starting as daemon, forking to background
[5][Sun Feb 16 17:04:38 2020][2143](src/main.c:254) MHD version is 0.9.51
[3][Sun Feb 16 17:04:38 2020][2143](src/main.c:270) libmicrohttpd is out of date, please upgrade to version 0.9.69 or higher
[3][Sun Feb 16 17:04:38 2020][2143](src/main.c:273) exiting...
pi@raspberrypi:~ $ 

Trying to upgrade libmicrohttpd shows I already have the latest version which is: "0.9.51-1"

To sum it up: nodogsplash won't start until I upgrade to a newer version of libmicrohttpd but it's already upgraded to latest version according to my installer.
Am I missing something or is this really an issue waiting to be resolved?

Hardware: Raspberry Pi 3 model B v1.2
System: Raspbian Stretch

**Comment by bluewavenet** [comment]
@Kosa103 
The latest release of libmicrohttpd (MHD) is 0.9.70
You can find it at:
https://ftp.gnu.org/gnu/libmicrohttpd/

You can force use of the old version at the risk of loss or corruption of form data,
for example when using passwords etc.
To do this you need to set the option in the nodogsplash.conf file:
`use_outdated_mhd 1`
 
For full details see the relevant section in the config file.

**Comment by Kosa103** [comment]
@bluewavenet 
Thank you! I enabled outdated versions as you said. Nodogsplash now starts as shown in terminal.

However the captive portal screen still doesn't show up. Something is going the way as intended because the AP doesn't provide internet access now (since I didn't pass the captive portal screen - but it just doesn't show up at all). 
I'm running raspap to create an AP. I use a wifi dongle on wlan1 instead of eth0 cable as a source of internet and the built-in wlan0 is set to AP mode - if these matter anything. Everything goes well without nodogsplash. "nodogsplash.conf" is set to "GatewayInterface wlan0" as it should be.

Can you help in this?

**Comment by bluewavenet** [comment]
@Kosa103 
What device are you using to connect to your wifi; mobile, laptop, desktop etc?

To see if the client device is running CPD, run the command:
`ndsctl status`
This should initially show your client device as preauthenticated.

I am not familiar with raspap, but I do know the terminology used in the rpi world is not accurate when it comes to routing. You need to be configured as a Wireless Router, not as a Wireless Access Point - can you confirm this is the case?

You might find the online documentation useful, although as you are most likely using the development snapshot (v4.4.1beta) of nodogsplash (NDS) and not a release version, the documentation might not be completely up to date.
https://nodogsplashdocs.readthedocs.io/en/latest/

Particularly read:
https://nodogsplashdocs.readthedocs.io/en/latest/howitworks.html

**Comment by bluewavenet** [comment]
@Kosa103 
I see you edited your last comment as i was writing my reply :-D

**Comment by bluewavenet** [comment]
@Kosa103 

> By directly accessing 10.3.141.1:2050 I can see the intended captive portal page but not when trying to connect any site the ordinary way. Pressing on "continue" first gets me a 404 error 

Yes - **_403 Forbidden_** - implying don't do this!
My guess is that the device you are using to connect is not running a CPD.
Try a mobile.
Or try a port 80 only web page eg:
http://onboard-wifi.net

**Comment by Kosa103** [comment]
@bluewavenet 
Thank you for your enthusiasm! I didn't expect this kind of helpfulness :)

I'm using a mobile phone to connect. Didn't have the chance to use any other device yet.
`ndsctl status` shows the device as preauthenticated! **Does this mean it's running CPD?** It's a Xiaomi redmi 6 with fairly new Android (MIUI Global 11.0.2).

status before passing the splash page:
![Nodogsplash Control - not connected](https://user-images.githubusercontent.com/61117806/74707224-793e2f80-5219-11ea-9af4-97a0535652ad.png)

status after passing splash page:
![Nodogsplash Control - connected](https://user-images.githubusercontent.com/61117806/74707249-8ce99600-5219-11ea-90d7-58a047d5c75b.png)

The raspap FAQ to set up a captive portal directly advises nodogsplash since they work well together, raspap not needing any config, nodogsplash needing only 2 little changes - which I made.
Here is the FAQ page I used:
[(https://github.com/billz/raspap-webgui/wiki/Captive-portal-setup)]

quoting the FAQ:
> The nodogsplash project is a lightweight, highly configurable captive portal solution. It integrates nicely with RaspAP and is recommended over other methods. No configuration changes are needed with RaspAP, however you will need to modify some default settings in the nodogsplash config.

I nonetheless checked what configuration I have and I found AP mode is disabled:
![Raspap AP disabled](https://user-images.githubusercontent.com/61117806/74707475-3b8dd680-521a-11ea-926b-aff5a2c72d87.png)

I'm connecting to the internet through a wifi dongle on wlan1 and using the pi's internal wifi for incoming connections (wlan0) so the captive portal is set up on wlan0.

I hope these settings are in place.
I tried opening the port 80 only page you wrote - http://onboard-wifi.net - still didn't work. Had to manually enter 10.3.141.1:2050 to bring up the splash page.

Based on what I saw and the minimal experience I have I assume the page does not get redirected to the captive portal (10.3.141.1:2050) for some reason but everything else seems to be fine. Can you help me resolving this?

**Comment by bluewavenet** [comment]
@Kosa103 
This is a different issue to your original problem. So I am closing this.
However we can continue here. Perhaps we can add some documentation about using raspap when we find the problem.
I don't have access to an rpi so it is up to you to try things :-D

**Comment by bluewavenet** [comment]
@Kosa103 
The links to raspap don't really go into any detail as far as I can see. So I will ask you some questions.

But first, here are some general prerequisites for getting NDS to work:
 * NDS must be installed on an IPv4 router
 * That router must have at least two network interfaces, WAN and LAN
 * one interface (WAN) must be connected to a live Internet feed and must be configured to obtain its IPv4 configuration via DHCP from whatever it is connected to.
 * The second interface (LAN) must have a static IPv4 address in an IPv4 subnetwork that is totally different the the subnetwork that the Internet feed is using.
 * The second (LAN) intervace IPv4 address must be the default IPv4 gateway for its subnet
 * The second (LAN) interface must be providing DHCP and DNS services to devices that connect to it.
 * The second (LAN) interface must provide network address translation (NAT) for all traffic that is forwarded to the first (WAN) interface.

**_Can you confirm your configuration meets ALL of these prerequisites?_**
If you are not sure of any of these, ask me for more detail.

**Comment by Kosa103** [comment]
Answers to your questions:


> - NDS must be installed on an IPv4 router 

-- **It is.** My RPi is receiving signal to WAN via WIFI-stick. It's using the built-in WIFI module to create a different WIFI network for other devices to connect. Works no problem without NDS


> - That router must have at least two network interfaces, WAN and LAN

-- **It does.**

> - one interface (WAN) must be connected to a live Internet feed and must be configured to obtain its IPv4 configuration via DHCP from whatever it is connected to.

-- **It is.** The wlan1 dongle does that. IP address is obtained from an outside router which provides access to WAN.

> - The second interface (LAN) must have a static IPv4 address in an IPv4 subnetwork that is totally different the the subnetwork that the Internet feed is using.

-- **It does.** IP address is 10.3.141.1 (RaspAP default)

> - The second (LAN) interface IPv4 address must be the default IPv4 gateway for its subnet

-- **I think so:** 10.3.141.1

> - The second (LAN) interface must be providing DHCP and DNS services to devices that connect to it.

-- **DHCP: Yes.** Connecting devices receive an IP address of 10.3.141.x (x is a number other than 1)
-- **DNS: NOT SURE! How do I check?**

> - The second (LAN) interface must provide network address translation (NAT) for all traffic that is forwarded to the first (WAN) interface.

-- **ALSO NOT SURE!** How do I check?


I'll provide you with an illustration of my network with IP addresses:

-------------------------------------------------------------------------------------------------------
![NETWORK](https://user-images.githubusercontent.com/61117806/75035184-36958500-54af-11ea-91bb-4990aa4e2c9a.png)
-------------------------------------------------------------------------------------------------------


I'm looking forward to your help! How to check above mentioned if DNS and NAT is provided?
And the original question - how can I make the captive portal appear automatically?
Many thanks you for your time! :)

**Comment by Kosa103** [comment]
@bluewavenet 
Also thought this information might help you get closer to what's happening:
I ran: `sudo systemctl status nodogsplash.service` 

and got this: 

------------------------------------------------------------------------------------------------------
![Status1](https://user-images.githubusercontent.com/61117806/75036317-dfdd7a80-54b1-11ea-9067-2f72863f21a3.png)
![Status2](https://user-images.githubusercontent.com/61117806/75036322-e2d86b00-54b1-11ea-8c97-3a75b976b531.png)

------------------------------------------------------------------------------------------------------


Outdated libmicrohttpd shouldn't be a problem as it'll only miss 2 special characters in form data - unrelated to my issue.

**Comment by bluewavenet** [comment]
@Kosa103 

> "DNS_PROBE_FINISHED_NO_INTERNET"

Where do you see this?

It sounds like you do not have a local DNS.
What do you get if you type the following command?:
`ping onboard-wifi.net`

**Comment by Kosa103** [comment]
@bluewavenet 
I see that on my phone - Google Chrome - trying to access google.com, after the timeout.

ping onboard-wifi.net - from my (non-rooted) phone terminal emulator:
`ping: unknown host onboard-wifi.net`

ping onboard-wifi.net - from RPi terminal:
```
PING onboard-wifi.net (46.32.240.41) 56(84) bytes of data.
64 bytes from 46.32.240.41 (46.32.240.41): icmp_seq=1 ttl=53 time=51.9 ms
64 bytes from 46.32.240.41 (46.32.240.41): icmp_seq=2 ttl=53 time=47.6 ms
64 bytes from 46.32.240.41 (46.32.240.41): icmp_seq=3 ttl=53 time=48.5 ms
```

**Comment by bluewavenet** [comment]
@Kosa103 
This means there is no local DNS. Without DNS, NDS will not work.
Is your DHCP being provided by dnamasq?
If so you need to configure the DNS part of it. (See the prerequisites I mentioned earlier)
Your DHCP is probably setting  an external DNS server and of course this will be blocked until the client authenticates, meaning the CPD will not pop up the splash page.

**Comment by Kosa103** [comment]
@bluewavenet 
Sure, thanks! I'll go check them out. If you can tell me which config files I need to check that would speed up the process. If not, no problem I'll find them anyways.

You know the strange thing is the raspap guide prefers NDS because they work so smoothly together. Claiming no need to change anything except 1 file: 
`sudo nano /etc/nodogsplash/nodogsplash.conf`

```
# GatewayInterface is not autodetected, has no default, and must be set here.
# Set GatewayInterface to the interface on your router
# that is to be managed by Nodogsplash.
# Typically br-lan for the wired and wireless lan.
#
GatewayInterface wlan0
#
# Parameter: GatewayAddress
# Default: Discovered from GatewayInterface
#
# This should be autodetected on an OpenWRT system, but if not:
# Set GatewayAddress to the IP address of the router on
# the GatewayInterface.  This is the address that the Nodogsplash
# server listens on.
GatewayAddress 10.3.141.1
```

which has not much to do with DNS I just thought to mention it because it's odd.

**Comment by bluewavenet** [comment]
@Kosa103 

> If you can tell me which config files I need to check

I don't know what you have installed, but you could start with /etc/dnsmasq.conf

> the raspap guide prefers NDS because they work so smoothly together. Claiming no need to change anything except 1 file

I read the guide and it seems to be rather simplistic and most likely just a cut and paste from other old Rpi tutorials. As you are finding, I suspect they did not actually test what the guide recommends.
Some considerable time ago - I think it was back when NDS v2 was released, a change was made, for security reasons, to block unauthenticated client access to external DNS.

Here is the relevant section of the nodogsplash.conf file
```
FirewallRuleSet preauthenticated-users {
# For preauthenticated users to resolve IP addresses in their
# initial request not using the router itself as a DNS server.
# Leave commented to help prevent DNS tunnelling
#  FirewallRule allow tcp port 53	
#  FirewallRule allow udp port 53
```

Try un-commenting the last two lines shown here, then restart NDS.

**Comment by Kosa103** [comment]
@bluewavenet 
**You did it!** That's what the error was!
I uncommented those firewallrule lines and now the captive portal screen shows up as soon as I enter the network. :D I'll contact raspap to update their tutorial with this addition. Thanks a lot for your invaluable help!!

**Comment by bluewavenet** [comment]
@Kosa103 
Uncommenting those rules allows it to work, but this is just a workaround for an incorrect configuration. RaspAp should really, when configuring as a router, be setting up its own dhcp/dns (probably using Dnsmasq). This will not only be more secure but also do DNS caching to speed up access and reduce traffic on the uplink.

**Comment by rimbosUY** [comment]
> @Kosa103
> The latest release of libmicrohttpd (MHD) is 0.9.70
> You can find it at:
> https://ftp.gnu.org/gnu/libmicrohttpd/
> 
> You can force use of the old version at the risk of loss or corruption of form data,
> for example when using passwords etc.
> To do this you need to set the option in the nodogsplash.conf file:
> `use_outdated_mhd 1`
> 
> For full details see the relevant section in the config file.

I am unable to proceed with the manual update of libmicrohttpd.
This is the process I am doing in my raspberry pi 👎 
wget https://ftp.gnu.org/gnu/libmicrohttpd/libmicrohttpd-0.9.70.tar.gz
tar -xzvf libmicrohttpd-0.9.70.tar.gz
./configure
sudo make
sudo make install


Maybe I am missing some step. Can you help me please ?

**Comment by bluewavenet** [comment]
@rimbosUY 
Courtesy of @MarByteBeep  in issue #476 

```
These are my build steps (for all you future NDS on RPi users out there):

Get https://ftp.gnu.org/gnu/libmicrohttpd/libmicrohttpd-0.9.69.tar.gz and unzip it in ~\libmicrohttpd

# get dependencies for libmicrohttpd build system
sudo apt-get install automake
sudo apt-get install texinfo

# build and install libmicrohttpd
cd ~\libmicrohttpd
chmod +x configure
sudo ./configure
sudo make
sudo make install

# update shared library links and cache
sudo /sbin/ldconfig -v

# build and install nodogsplash
cd ~
git clone https://github.com/nodogsplash/nodogsplash.git
cd ~/nodogsplash
make
sudo make install
```

**Comment by martignoni** [comment]
@bluewavenet Much useful. Thank you.

**Comment by rimbosUY** [comment]
Thanks a lot @martignoni  and @bluewavenet for your fast response.

I have tried step by step but it didn't work. I continue getting the MHD version is 0.9.62
[3][Fri Mar 13 16:52:49 2020][32309](src/main.c:283) libmicrohttpd is out of date, please upgrade to version 0.9.69 or higher

Maybe it is because I have previously installed nodogsplash, so reinstalling it is not enough because it may be using some library from the cache.

I will plan to reinstall all the raspberry from scratch this weekend, so I can validate if this steps are the expected ones to update the libmicrohttpd library version.

Thanks again!

**Comment by bluewavenet** [comment]
@rimbosUY 
Did you reboot?

**Comment by rimbosUY** [comment]
yes. But it still continue with the same error message.
Maybe these logs give you any idea about what's wrong on my setup.

pi@raspberrypi:~ $ **sudo ldconfig -v | grep libmicro**
ldconfig: Can't stat /usr/local/lib/arm-linux-gnueabihf: No such file or directory
ldconfig: Path `/usr/local/lib' given more than once
ldconfig: Path `/lib/arm-linux-gnueabihf' given more than once
ldconfig: Path `/usr/lib/arm-linux-gnueabihf' given more than once
ldconfig: /lib/arm-linux-gnueabihf/ld-2.28.so is the dynamic linker, ignoring

	libmicrohttpd.so.12 -> libmicrohttpd.so.12.49.0
	libmicrohttpd.so.12 -> libmicrohttpd.so.12.55.0
ldconfig: /lib/ld-linux.so.3 is the dynamic linker, ignoring

pi@raspberrypi:~ $ **cd /usr/local/lib**
pi@raspberrypi:/usr/local/lib $ **ls**
libmicrohttpd.a   libmicrohttpd.so     libmicrohttpd.so.12.55.0  python2.7
libmicrohttpd.la  libmicrohttpd.so.12  pkgconfig                 python3.7
pi@raspberrypi:/usr/local/lib $

**Comment by martignoni** [comment]
Same here: after compiling using above instructions re-installing and rebooting, I'm still getting the same error.

**Comment by bluewavenet** [comment]
@martignoni @rimbosUY 
Interesting. I don't have a pi to test this on.
@MarByteBeep any ideas?

**Comment by MarByteBeep** [comment]
Hmmm. I can try to test this tomorrow. But that installation file worked for me. You are 100% sure you executed the 

sudo /sbin/ldconfig -v

Command?

**Comment by bluewavenet** [comment]
@MarByteBeep 
Much appreciated if you can retest ;-)
